osquery/events/darwin/endpointsecurity.cpp

```diff
@@ -0,0 +1,232 @@
+    default:
+      return Status::failure(1, "EndpointSecurity: Unknown error");
+    }
+  } else {
+    return Status::failure(
+        1, "EndpointSecurity is only available on macOS 10.15 and higher");
+  }
+}
+
+void EndpointSecurityPublisher::configure() {
+  if (es_client_ == nullptr) {
+    return;
+  }
+
+  auto cache = es_clear_cache(es_client_);
+  if (cache != ES_CLEAR_CACHE_RESULT_SUCCESS) {
+    VLOG(1) << "Couldn't clear cache for EndpointSecurity client";
+    return;
+  }
+
+  for (auto& sub : subscriptions_) {
+    auto sc = getSubscriptionContext(sub->context);
+    auto events = sc->es_event_subscriptions_;
+    auto es_sub = es_subscribe(es_client_, &events[0], events.size());
+    if (es_sub != ES_RETURN_SUCCESS) {
+      VLOG(1) << "Couldn't subscribe to EndpointSecurity subsystem";
+    }
+  }
+}
+
+void EndpointSecurityPublisher::tearDown() {
+  if (es_client_ == nullptr) {
+    return;
+  }
+  es_unsubscribe_all(es_client_);
+
+  if (es_client_success_) {
+    es_delete_client(es_client_);
+    es_client_ = nullptr;
+  }
+}
+
+static inline std::string getPath(const es_process_t* p) {
+  return p->executable->path.length > 0 ? p->executable->path.data : "";
+}
+
+static inline std::string getSigningId(const es_process_t* p) {
+  return p->signing_id.length > 0 ? p->signing_id.data : "";
+}
+
+static inline std::string getTeamId(const es_process_t* p) {
+  return p->team_id.length > 0 ? p->team_id.data : "";
+}
+
+static inline std::string getStringFromToken(es_string_token_t* t) {
+  return t->length > 0 && t->data != nullptr ? t->data : "";
+}
+
+static inline std::string getCwdPathFromPid(pid_t pid) {
+  struct proc_vnodepathinfo vpi {};
+  auto bytes = proc_pidinfo(pid, PROC_PIDVNODEPATHINFO, 0, &vpi, sizeof(vpi));
+  return bytes <= 0 ? "" : vpi.pvi_cdir.vip_path;
+}
+
+static inline std::string getCDHash(const es_process_t* p) {
+  std::stringstream hash;
+  for (unsigned char i : p->cdhash) {
+    hash << std::hex << std::setfill('0') << std::setw(2)
+         << static_cast<unsigned int>(i);
+  }
+  auto s = hash.str();
+  return s.find_first_not_of(s.front()) == std::string::npos ? "" : s;
+}
+
+static inline void getProperties(const es_process_t* p,
+                                 const EndpointSecurityEventContextRef& ec) {
+  auto audit_token = p->audit_token;
+  ec->pid = audit_token_to_pid(audit_token);
+  ec->parent = p->ppid;
+  ec->original_parent = p->original_ppid;
+
+  ec->path = getPath(p);
+  ec->cwd = getCwdPathFromPid(ec->pid);
+
+  ec->uid = audit_token_to_ruid(audit_token);
+  ec->euid = audit_token_to_egid(audit_token);
+  ec->gid = audit_token_to_rgid(audit_token);
+  ec->egid = audit_token_to_egid(audit_token);
+
+  ec->signing_id = getSigningId(p);
+  ec->team_id = getTeamId(p);
+  ec->cdhash = getCDHash(p);
+  ec->platform_binary = p->is_platform_binary;
+
+  auto user = getpwuid(ec->uid);
+  ec->username = std::string(user->pw_name);
```

Comment:
IIRC sometimes `user->pw_name` is a null pointer. I recommend adding a check before creating the `std::string`.

Reply:
Good catch, thanks!